addons/l10n_latam_document/models/account_invoice.py

```diff
@@ -0,0 +1,303 @@
+            rec.move_id.l10n_latam_document_type_id = \
+                rec.l10n_latam_document_type_id.id,
+        return True
+
+    @api.onchange('journal_id', 'partner_id', 'company_id')
+    def onchange_journal_partner_company(self):
+        res = self._get_available_document_types(
+            self.journal_id, self.type, self.partner_id)
+        self.l10n_latam_document_type_id = res['document_type']
+
+    @api.depends('journal_id', 'partner_id', 'company_id')
+    def _compute_l10n_latam_available_document_types(self):
+        """
+        This function should only be inherited if you need to add another
+        "depends", for eg, if you need to add a depend on "new_field" you
+        should add:
+
+        @api.depends('new_field')
+        def _get_available_document_types(
+                self, journal, invoice_type, partner):
+            return super(
+                AccountInvoice, self)._get_available_document_types(
+                    journal, invoice_type, partner)
+        """
+        for invoice in self:
+            res = invoice._get_available_document_types(
+                invoice.journal_id, invoice.type, invoice.partner_id)
+            invoice.l10n_latam_available_document_type_ids = res[
+                'available_document_types']
+
+    @api.multi
+    def write(self, vals):
+        """ If someone change the type (for eg from sale order), we update
+        the document type"""
+        inv_type = vals.get('type')
+        # if len(vals) == 1 and vals.get('type'):
+        # podrian pasarse otras cosas ademas del type
+        if not inv_type:
+            return super(AccountInvoice, self).write(vals)
+
+        for rec in self:
+            res = rec._get_available_document_types(
+                rec.journal_id, inv_type, rec.partner_id)
+            vals['l10n_latam_document_type_id'] = res['document_type'].id
+            # call write for each invoice
+            super(AccountInvoice, rec).write(vals)
+        return True
+
+    @api.model
+    def _get_available_document_types(
+            self, journal, invoice_type, partner):
+        """ This function is to be inherited by different localizations and
+        MUST return a dictionary with two keys:
+        * 'available_document_types': available document types on
+        this invoice
+        * 'document_type': suggested document type on this invoice
+        """
+        if not journal.l10n_latam_use_documents:
+            return {
+                'available_document_types':
+                    self.env['l10n_latam.document.type'],
+                'document_type':
+                    self.env['l10n_latam.document.type'],
+            }
+        raise UserError(_(
+            'Method not implemented by localization of %s') % (
+                journal.company_id.country_id.name))
+
+    @api.constrains('type', 'l10n_latam_document_type_id')
+    def check_invoice_type_document_type(self):
```

Comment:
Yes, on an ideal word this shouldn't happend, but with the constraint we are sure to avoid errors.

For eg, we have removed the "write" definition on l10n_latam_documents.invoice and this constraint help us to detect that there was an incosistent invoice created when refunding form sale orders.

We think better to keep the constraint to ensure everthing goes ok